File: ios/Packages/AppFeatures/Sources/DuaFeature/DuaViewModel.swift
import CoreKit
import ContentKit
import Foundation
import Observation

/// Library browse/search/favorite state (docs/features/dua.md). Pure over
/// injected store/clock for favorites; category loading from ContentKit is a
/// thin wrapper kept separate so search/favorite logic is unit-testable with
/// hand-built fixtures.
@MainActor
@Observable
public final class DuaViewModel {
    public private(set) var categories: [DuaCategory] = []
    public private(set) var isLoadingCategories = false
    public var searchQuery: String = "" {
        didSet { recordSearchIfNeeded() }
    }

    /// duaId -> addedAt. A dictionary (not an array-index) so favorites survive
    /// content resync — the spec's "keyed by stable duaId" requirement.
    private var favorites: [String: Date] = [:]

    private let contentService: ContentService?
    private let store: DuaStoring
    private let haptics: HapticsProviding
    private let now: @Sendable () -> Date
    private let searchHistory: SearchHistoryRecording
    private var locale = "ar"
    /// Dedupes recording on every keystroke that still resolves to the same
    /// completed query (e.g. focus loss re-triggering) — not a full debounce,
    /// but the spec explicitly allows either.
    private var lastRecordedQuery: String?

    public init(
        contentService: ContentService? = nil,
        store: DuaStoring,
        haptics: HapticsProviding = NoopHaptics(),
        now: @escaping @Sendable () -> Date = { Date() },
        searchHistory: SearchHistoryRecording = NoopSearchHistoryRecording()
    ) {
        self.contentService = contentService
        self.store = store
        self.haptics = haptics
        self.now = now
        self.searchHistory = searchHistory
        self.favorites = Dictionary(uniqueKeysWithValues: store.loadFavorites().map { ($0.duaId, $0.addedAt) })
    }

    public func loadCategories(locale: String) async {
        self.locale = locale
        guard let contentService else { return }
        isLoadingCategories = true
        setCategories(await contentService.duas(locale: locale)?.categories ?? [])
        isLoadingCategories = false
    }

    private func recordSearchIfNeeded() {
        guard let results = searchResults, !results.isEmpty, searchQuery != lastRecordedQuery else { return }
        lastRecordedQuery = searchQuery
        searchHistory.record(source: "dua", queryText: searchQuery, locale: locale)
    }

    /// Internal seam so search/favorite logic is testable without a live
    /// ContentService (mirrors AzkarViewModel.startSession taking items directly).
    func setCategories(_ categories: [DuaCategory]) {
        self.categories = categories
    }

    public var allDuas: [Dua] { categories.flatMap(\.duas) }

    /// Most-recently-favorited first.
    public var favoriteDuas: [Dua] {
        allDuas
            .filter { favorites[$0.id] != nil }
            .sorted { favorites[$0.id]! > favorites[$1.id]! }
    }

    public func isFavorite(_ duaId: String) -> Bool {
        favorites[duaId] != nil
    }

    public func toggleFavorite(_ duaId: String) {
        if favorites[duaId] != nil {
            favorites.removeValue(forKey: duaId)
            haptics.tick()
        } else {
            favorites[duaId] = now()
            haptics.targetReached()
        }
        store.saveFavorites(favorites.map { FavoriteDua(duaId: $0.key, addedAt: $0.value) })
    }

    /// `nil` means "not searching" (empty query); an empty array means
    /// "searched, no matches" — screens must distinguish these (spec §3).
    public var searchResults: [Dua]? {
        let normalized = DuaSearch.normalize(searchQuery)
        guard !normalized.isEmpty else { return nil }
        return allDuas.filter { dua in
            DuaSearch.normalize(dua.title).contains(normalized) ||
                DuaSearch.normalize(dua.arabicText).contains(normalized) ||
                (dua.translation.map { DuaSearch.normalize($0).contains(normalized) } ?? false)
        }
    }
}

File: android/feature/dua/src/main/kotlin/com/fatwabot/feature/dua/DuaViewModel.kt
package com.fatwabot.feature.dua

import androidx.lifecycle.ViewModel
import com.fatwabot.core.common.HapticsProviding
import com.fatwabot.core.common.NoopSearchHistoryRecording
import com.fatwabot.core.common.SearchHistoryRecording
import com.fatwabot.core.content.ContentService
import com.fatwabot.core.content.Dua
import com.fatwabot.core.content.DuaCategory
import dagger.hilt.android.lifecycle.HiltViewModel
import javax.inject.Inject
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.datetime.Clock

/**
 * Library browse/search/favorite state — mirror of iOS DuaViewModel
 * (docs/features/dua.md). Category loading from ContentKit is a thin
 * synchronous wrapper; search/favorite logic is unit-testable with
 * hand-built fixtures via [setCategories].
 */
@HiltViewModel
class DuaViewModel @Inject constructor(
    private val contentService: ContentService?,
    private val store: DuaStoring,
    private val clock: Clock,
    private val haptics: HapticsProviding,
    private val searchHistory: SearchHistoryRecording = NoopSearchHistoryRecording(),
) : ViewModel() {
    private var locale = "ar"
    /** Dedupes recording on repeated identical queries (e.g. focus loss
     * re-triggering) — not a full debounce, but the spec explicitly allows either. */
    private var lastRecordedQuery: String? = null

    data class UiState(
        val categories: List<DuaCategory> = emptyList(),
        val hasLoadedCategories: Boolean = false,
        val searchQuery: String = "",
        /** duaId -> addedAt epoch seconds. Dictionary (not array-index) so
         * favorites survive content resync — the spec's "keyed by stable
         * duaId" requirement. */
        val favorites: Map<String, Long> = emptyMap(),
    ) {
        val allDuas: List<Dua>
            get() = categories.flatMap { it.duas }

        /** Most-recently-favorited first. */
        val favoriteDuas: List<Dua>
            get() = allDuas.filter { favorites.containsKey(it.id) }
                .sortedByDescending { favorites[it.id] }

        fun isFavorite(duaId: String): Boolean = favorites.containsKey(duaId)

        /** `null` means "not searching" (empty query); an empty list means
         * "searched, no matches" — screens must distinguish these. */
        val searchResults: List<Dua>?
            get() {
                val normalized = DuaSearch.normalize(searchQuery)
                if (normalized.isEmpty()) return null
                return allDuas.filter { dua ->
                    DuaSearch.normalize(dua.title).contains(normalized) ||
                        DuaSearch.normalize(dua.arabicText).contains(normalized) ||
                        (dua.translation?.let { DuaSearch.normalize(it).contains(normalized) } ?: false)
                }
            }
    }

    private val _state = MutableStateFlow(
        UiState(
            favorites = store.loadFavorites().associate { it.duaId to it.addedAtEpochSeconds },
        ),
    )
    val state: StateFlow<UiState> = _state.asStateFlow()

    fun loadCategories(locale: String) {
        this.locale = locale
        setCategories(contentService?.duas(locale)?.categories.orEmpty())
        _state.update { it.copy(hasLoadedCategories = true) }
    }

    /** Internal seam so search/favorite logic is testable without a live
     * ContentService (mirrors AzkarViewModel.startSession taking items directly). */
    fun setCategories(categories: List<DuaCategory>) {
        _state.update { it.copy(categories = categories) }
    }

    fun updateSearchQuery(query: String) {
        _state.update { it.copy(searchQuery = query) }
        val results = _state.value.searchResults
        if (results != null && results.isNotEmpty() && query != lastRecordedQuery) {
            lastRecordedQuery = query
            searchHistory.record(source = "dua", queryText = query, locale = locale)
        }
    }

    fun toggleFavorite(duaId: String) {
        val current = _state.value.favorites
        val updated = if (current.containsKey(duaId)) {
            haptics.tick()
            current - duaId
        } else {
            haptics.targetReached()
            current + (duaId to clock.now().epochSeconds)
        }
        store.saveFavorites(updated.map { FavoriteDua(it.key, it.value) })
        _state.update { it.copy(favorites = updated) }
    }
}
